python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,320 @@
+        .orElse(null),
+      parsedParameters);
+  }
+
+  private record PipelineNameAndParsedParameters(Name pipelineName, Map<String, Set<StringAndTree>> parsedParameters) {
+  }
+
+  private static Optional<PipelineNameAndParsedParameters> newPipelineNameAndParsedParameters(@Nullable Name pipelineName, Map<String, Set<StringAndTree>> parsedParameters) {
+    return Optional.ofNullable(pipelineName).map(pipelineName1 -> new PipelineNameAndParsedParameters(pipelineName1, parsedParameters));
+  }
+
+  private static void findProblems(Map<String, Set<StringAndTree>> setParameters, Map<String, ClassSymbol> pipelineDefinition, SubscriptionContext subscriptionContext) {
+    for (var entry : setParameters.entrySet()) {
+      var step = entry.getKey();
+      var stringAndTree = entry.getValue();
+      var parameters = stringAndTree.stream().map(StringAndTree::string).collect(Collectors.toSet());
+
+      var classifier = pipelineDefinition.get(step);
+      if (classifier == null) {
+        continue;
+      }
+      var possibleParameters = getInitFunctionSymbol(classifier).map(FunctionSymbol::parameters).orElse(List.of());
+
+      parameters.forEach(parameter -> {
+        if (isNotAValidParameter(parameter, possibleParameters)) {
+          createIssue(subscriptionContext, parameter, stringAndTree);
+        }
+      });
+    }
+  }
+
+  private static void createIssue(SubscriptionContext subscriptionContext, String parameter, Set<StringAndTree> stringAndTree) {
+    stringAndTree.stream().filter(stringAndTree1 -> stringAndTree1.string().equals(parameter)).findFirst()
+      .ifPresent(location -> subscriptionContext.addIssue(location.tree, MESSAGE));
+  }
+
+  private static boolean isNotAValidParameter(String parameter, List<FunctionSymbol.Parameter> possibleParameters) {
+    return possibleParameters.stream().noneMatch(symbol -> Objects.equals(symbol.name(), parameter));
+  }
+
+  private static Optional<FunctionSymbol> getInitFunctionSymbol(ClassSymbol classSymbol) {
+    return classSymbol.declaredMembers().stream().filter(
+      memberSymbol -> "__init__".equals(memberSymbol.name())).findFirst().map(FunctionSymbol.class::cast);
+  }
+
+  private static Stream<Expression> getExpressionsFromArgument(@Nullable RegularArgument argument) {
+    return Optional.ofNullable(argument)
+      .map(RegularArgument::expression)
+      .flatMap(TreeUtils.toOptionalInstanceOfMapper(ListLiteralImpl.class))
+      .map(ListLiteralImpl::elements)
+      .map(ExpressionList::expressions)
+      .stream()
+      .flatMap(Collection::stream);
+  }
+
+  private record ExpressionAndPrefix(List<Expression> tuple, String prefix) {
+  }
+
+  private static Map<String, ClassSymbol> parsePipeline(CallExpression callExpression) {
+    var stepsArgument = TreeUtils.nthArgumentOrKeyword(0, "steps", callExpression.arguments());
+    var out = new HashMap<String, ClassSymbol>();
+
+    var a = getExpressionsFromArgument(stepsArgument)
```

Comment:
Here I guess the `a` could be removed (and the `forEach` appended to the stream) as it does not help with the comprehension of the code.

Reply:
Good catch, it was a remnant from the debugging phase